### PR TITLE
Tactic FSM Framework

### LINKED ISSUE #1895: Create FSM Tactic framework
### Description of the task
Create a framework for adding FSMs to Tactics with some example Tactics

### Acceptance criteria

- [ ] Should not break existing interfaces so that transition to FSMs can be done gradually
- [ ] convert some basic tactics

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Created framework for adding FSMs to Tactics
* Made example Move and Stop Tactics which are used in the simualated example play test
* preserves the current Tactic interface while adding support for FSM interface
* modified the Play interface to fix
* Some refactoring to Tactic to make it work
* Had to modify playinfo logic and fixed a bug

This means that tactics can be gradually transitioned into the new system without having to halt any existing development

### Testing Done
regression tests, unit tests

### Resolved Issues
resolves #1895
part of #1831

### Length Justification
a lot of it is just renaming and refactoring

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/move_tactic.h [hunk 1]
```diff
@@ -1,7 +1,55 @@
...
+                DribblerMode::OFF, BallCollisionType::AVOID));
+        };
+
+        const auto movement_done = [](auto event) {
+            return moveRobotDone(event.common.robot, event.controls.destination,
+                                 event.controls.final_orientation);
+        };
+
+        return make_transition_table(
+            *"idle"_s + event<Update> / update_move_intent = state<Move>,
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/tactic/move_tactic.h]: since we always call `update_move_intent` seems like this could just be a one-liner?

`X + event<Update> / update_move_intent = X)`

**Review by MathewMacDougall — COMMENTED** [review]
Overall the FSM stuff looks pretty clean. Glad that the TacticAssignment stuff injected into Plays doesn't affect how Plays are written :+1: 

Main thing I would really like to see is how this looks with a _much_ more complicated example. Eg. a Tactic that intercepts the ball, dribbles to a destination, and shoots (obviously actual gameplay logic can be mocked for this, but required states should be fully fleshed out). I think that would give a much better idea how this will look and deal with complexity.

More complex example also helps answer:
1. How to composite FSMs look in general? Are they are easy to follow as we think across multiple levels?
2. How are child FSMs coordinated by the parent Tactic? (See my comments in the google doc).

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: X means that the Tactic is done so that's why there's extra logic here

File: src/software/ai/hl/stp/tactic/tactic.h [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+// TODO (#1888): remove this typedef
 // We typedef the coroutine return type to make it shorter, more descriptive,
 // and easier to work with
 typedef boost::coroutines2::coroutine<std::shared_ptr<Action>> ActionCoroutine;
 
+struct TacticUpdate
+{
+    Robot robot;
+    World world;
+    std::function<void(std::unique_ptr<Intent>)> set_intent;
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/tactic.h]: Would be nice if there was comment here on what `set_intent` should be

File: src/software/ai/hl/stp/tactic/tactic.cpp [hunk 3]
```diff
@@ -107,3 +109,37 @@ std::set<RobotCapability> &Tactic::mutableRobotCapabilityRequirements()
...
+{
+    // TODO (#1888): remove these updates
+    updateWorldParams(world);
+    updateRobot(robot);
+
+    updateIntent(TacticUpdate{.robot      = robot,
+                              .world      = world,
+                              .set_intent = [this](std::unique_ptr<Intent> new_intent) {
+                                  intent = std::move(new_intent);
+                              }});
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/tactic.cpp]: why does this function need to be dependency injected on every get?

is this mostly to capture "this" tactic and set the intent?

File: src/software/ai/hl/stp/tactic/tactic.cpp [hunk 4]
```diff
@@ -107,3 +109,37 @@ std::set<RobotCapability> &Tactic::mutableRobotCapabilityRequirements()
...
+    if (intent)
+    {
+        return std::move(intent);
+    }
+    else
+    {
+        LOG(WARNING) << "No intent set for this tactic" << std::endl;
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/tactic.cpp]: This is in `tactic.cpp` we won't know which tactic this is, maybe worth adding `TYPENAME(*this)` here (although not sure if that will even work)?

File: src/software/ai/hl/stp/tactic/stop_tactic.h [hunk 5]
```diff
@@ -1,7 +1,41 @@
...
+    };
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto update_stop_intent = [](auto event) {
+            event.common.set_intent(
+                std::make_unique<StopIntent>(event.common.robot.id(), event.coast));
+        };
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/stop_tactic.h]: Does this function need to be defined in every tactic that wants to use the `stop_intent`?

From what I understand the event is of type `StopFSM::Update`,  so it seems like every tactic is basically going to have the same copy of these `update_.*_intent` actions for their own `.*FSM::Update` struct. 

Is there anyway we can avoid this?

File: src/software/ai/hl/stp/tactic/stop_tactic.h [hunk 6]
```diff
@@ -1,7 +1,41 @@
...
+
+        const auto stop_done = [](auto event) {
+            return robotStopped(event.common.robot);
+        };
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/stop_tactic.h]: ditto above: its great the `robotStopped` is shared, but can we also share this guard? almost all tactics will probably want this exact same guard.

File: src/software/ai/hl/stp/tactic/stop_tactic.h [hunk 7]
```diff
@@ -1,7 +1,41 @@
...
-#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/move_action.h"  // TODO (#1888): remove this dependency
 #include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/stop_intent.h"
+
+struct StopFSM
+{
+    struct Stop
+    { /* state */
+    };
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/stop_tactic.h]: Is this preferred over just using "Stop"_s (similar to "idle"_s)?

File: src/software/ai/hl/stp/tactic/tactic.cpp [hunk 8]
```diff
@@ -107,3 +109,37 @@ std::set<RobotCapability> &Tactic::mutableRobotCapabilityRequirements()
...
+        return std::make_unique<StopIntent>(robot.id(), false);
+    }
+}
+
+void Tactic::updateIntent(const TacticUpdate &tactic_update)
+{
+    // Try to get an intent from the tactic
+    std::shared_ptr<Action> action = getNextAction();
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/tactic.cpp:138]: Is this action stuff just to not break compatibility w/ other tactics? will this function will go away with #1888?

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Exciting stuff! can't wait to use the FSMs

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 1 above)
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/move_tactic.h]: This table will look at confusing to new members. If this is used as the reference for FSM tactics, it would be good to label the columns of the table with src state, event, guard, and dest state.

File: src/software/ai/hl/stp/tactic/move_tactic.h [hunk 9]
```diff
@@ -1,7 +1,55 @@
 #pragma once
 
-#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/move_action.h"  // TODO (#1888): remove this dependency
 #include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/move_intent.h"
+
+struct MoveFSM
+{
+    struct Move
```
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/move_tactic.h]: It would be good to have comment lines to separate state structs from other structs.

File: src/software/ai/hl/stp/tactic/transition_conditions.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @param robot The robot to check
+ * @param destination The destination to move to
+ * @param final_orientation The final orientation of the movement
+ * @param DISTANCE_THRESHOLD distance from destination to be considered close enough
+ * @param ANGLE_THRESHOLD angle from final orientation to be considered close enough
+ *
+ * @return true if robot movement is done
+ */
+bool moveRobotDone(const Robot& robot, const Point& destination,
```
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/transition_conditions.h]: nit: this function name doesn't make it clear what the function does, maybe something like `robotReachedDestination` would be clearer

File: src/software/ai/hl/stp/tactic/move_tactic.h [hunk 11]
```diff
@@ -1,7 +1,55 @@
...
+    { /* state */
+    };
+
+    struct Controls
```
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/move_tactic.h]: This name could be confusing, I suggest `Parameters`

**Review by jordanzliu — CHANGES_REQUESTED** [review]
Good work, I really like the idea of having FSM tactics! Just a few minor stylistic and readability issues

File: src/software/ai/hl/stp/tactic/tactic.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic.cpp]: Not sure what you mean, set_intent should be a member of Tactic?

File: src/software/ai/hl/stp/tactic/stop_tactic.h (hunk 5 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/stop_tactic.h]: I don't think there's anything we can do here

File: src/software/ai/hl/stp/tactic/stop_tactic.h (hunk 6 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/stop_tactic.h]: since the event is different, it's not possible to share this guard

File: src/software/ai/hl/stp/tactic/stop_tactic.h (hunk 7 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/stop_tactic.h]: Yes, because it prevents typos

File: src/software/ai/hl/stp/tactic/tactic.cpp (hunk 8 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic.cpp:138]: yes

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 9 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: done

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: done

File: src/software/ai/hl/stp/tactic/transition_conditions.h (hunk 10 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/transition_conditions.h]: done

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 11 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: renamed to controlparams

File: src/software/ai/hl/stp/play/play.h [hunk 12]
```diff
@@ -1,12 +1,20 @@
...
 
+#include <boost/bind.hpp>
 #include <boost/coroutine2/all.hpp>
 #include <memory>
 #include <vector>
 
 #include "software/ai/hl/stp/tactic/tactic.h"
 #include "software/world/world.h"
 
+using RobotToTacticAssignmentAlgorithm =
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.h]: renamed to RobotToTacticAssignmentFunction

File: src/software/ai/hl/stp/play/play.cpp [hunk 13]
```diff
@@ -39,6 +39,38 @@ std::vector<std::shared_ptr<Tactic>> Play::getTactics(const World &world)
...
+    for (const auto tactic : tactics)
+    {
+        const_tactics.push_back(tactic);
+    }
+    return const_tactics;
+}
+
+std::vector<std::unique_ptr<Intent>> Play::get(
+    RobotToTacticAssignmentAlgorithm robot_to_tactic_assignment_algorithm,
+    MotionConstraintCalculator motion_constraint_builder, const World &new_world)
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.cpp]: renamed to MotionConstraintBuildFunction

File: src/software/ai/hl/stp/tactic/tactic.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic.h]: done

File: src/software/ai/hl/stp/tactic/tactic.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic.cpp]: done

File: src/software/ai/hl/stp/play/example_play_test.cpp [hunk 14]
```diff
@@ -1,41 +0,0 @@
-#include "software/ai/hl/stp/play/example_play.h"
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play/example_play_test.cpp:1]: This might be elsewhere, but showing how plays will be tested seems like a pretty key point (might just be sim tests, which seems like a reasonable approach to me).

File: src/software/ai/hl/stp/play/play.cpp [hunk 15]
```diff
@@ -39,6 +39,31 @@ std::vector<std::shared_ptr<Tactic>> Play::getTactics(const World &world)
...
+std::vector<std::unique_ptr<Intent>> Play::get(
+    RobotToTacticAssignmentFunction robot_to_tactic_assignment_algorithm,
+    MotionConstraintBuildFunction motion_constraint_builder, const World &new_world)
+{
+    std::vector<std::unique_ptr<Intent>> intents;
+    auto tactics = getTactics(new_world);
+    std::vector<std::shared_ptr<const Tactic>> const_tactics;
+    const_tactics.reserve(tactics.size());
+    std::transform(tactics.begin(), tactics.end(), std::back_inserter(const_tactics),
+                   [](std::shared_ptr<Tactic> tactic) { return tactic; });
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play/play.cpp:52]: Probably worth adding a comment here explaining why we need this `std::transform`.

File: src/software/ai/hl/stp/play/play.cpp [hunk 16]
```diff
@@ -39,6 +39,31 @@ std::vector<std::shared_ptr<Tactic>> Play::getTactics(const World &world)
...
+        robot_to_tactic_assignment_algorithm(const_tactics, new_world);
+    for (auto tactic : tactics)
+    {
+        auto iter = robot_tactic_assignment.find(tactic);
+        if (iter != robot_tactic_assignment.end())
+        {
+            auto intent = tactic->get(iter->second, new_world);
+            intent->setMotionConstraints(motion_constraint_builder(*tactic));
+            intents.push_back(std::move(intent));
+        }
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play/play.cpp:63]: We should probably log at least a warning if the robot doesn't exist.

File: src/software/ai/hl/stp/play/play.h [hunk 17]
```diff
@@ -66,6 +74,32 @@ class Play
...
+     * Gets Intents from the Play given the assignment algorithm and world
+     *
+     * @param robot_to_tactic_assignment_algorithm The algorithm for assigning robots to
+     * tactics
+     * @param motion_constraint_builder Builds motion constraints from tactics
+     * @param world The updated world
+     *
+     * @return the vector of intents to execute
+     */
+    std::vector<std::unique_ptr<Intent>> get(
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play/play.h:96]: nit; but `getIntents` might be a bit more descriptive if a user is just reading a call to this function without the jdoc in front of them.

File: src/software/ai/hl/stp/play_info.cpp [hunk 18]
```diff
@@ -1,7 +1,7 @@
 #include "software/ai/hl/stp/play_info.h"
 
 PlayInfo::PlayInfo(std::string referee_command_name, std::string play_name,
-                   std::unordered_set<std::string> robot_tactic_assignment)
+                   std::vector<std::string> robot_tactic_assignment)
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play_info.cpp:4]: Did order become important for this? Why the change?

File: src/software/ai/hl/stp/play_info.h [hunk 19]
```diff
@@ -57,5 +56,5 @@ class PlayInfo
    private:
     std::string referee_command_name;
     std::string play_name;
-    std::unordered_set<std::string> robot_tactic_assignment;
+    std::vector<std::string> robot_tactic_assignment;
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play_info.h:59]: Similar comment here. Did the order become important?

File: src/software/ai/hl/stp/stp.cpp [hunk 20]
```diff
@@ -21,8 +22,15 @@
...
+      current_play(nullptr),
+      readable_robot_tactic_assignment(),
       random_number_generator(random_seed),
-      control_config(control_config)
+      control_config(control_config),
+      override_play_name(""),
+      previous_override_play_name(""),
+      override_play(false),
+      previous_override_play(false),
+      current_game_state()
```
Comment (garethellis0) [inline src/software/ai/hl/stp/stp.cpp:33]: Might be worth adding some sort of abstraction for all the overriding stuff, seems like this class is getting a bit bloated.

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 1 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: Do you do this in every tactic though? One option that I've used as a middle ground on a few student teams is to have 1 instance of each class that's really heavily commented so that you can direct new people to look at that as a reference without having to duplicate comments.

File: src/software/ai/hl/stp/tactic/move_tactic.h [hunk 21]
```diff
@@ -1,7 +1,59 @@
 #pragma once
 
-#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/move_action.h"  // TODO (#1888): remove this dependency
 #include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/move_intent.h"
+
+struct MoveFSM
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/move_tactic.h]: I might be missing a trick, but does this need to be publicly exposed? Seems like it's just an implementation detail for this tactic. If so, might be worth defining in an anonymous namespace.

File: src/software/ai/hl/stp/tactic/move_tactic.h [hunk 22]
```diff
@@ -1,7 +1,59 @@
 #pragma once
 
-#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/move_action.h"  // TODO (#1888): remove this dependency
 #include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/move_intent.h"
+
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/move_tactic.h:7]: I really like this incremental migration approach. :ok_hand:

File: src/software/ai/hl/stp/tactic/tactic.h [hunk 23]
```diff
@@ -121,10 +151,12 @@ class Tactic
     virtual ~Tactic() = default;
 
    protected:
+    // TODO (#1888): remove this field
     // The robot performing this Tactic
-    std::optional<Robot> robot;
+    std::optional<Robot> robot_;
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/tactic.h:147]: As a general comment, I would strongly recommend using `.value()` when you're accessing `std::optional` values instead of `*` or `->` (I'm sure a past me is guilty of doing this in the codebase). This is because `.value()` actually checks the optional is valid, so you immediately get a segfault with a (hopefully) comprehensible error, whereas using the other two options will not check anything, and so the error might not show up until much later.

File: src/software/ai/hl/stp/tactic/stop_tactic.h [hunk 24]
```diff
@@ -1,7 +1,43 @@
 #pragma once
 
-#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/move_action.h"  // TODO (#1888): remove this dependency
 #include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/stop_intent.h"
+
+struct StopFSM
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/stop_tactic.h]: similar comment here about this being exposed publicly, consider putting it in anonymous namespace.

File: src/software/ai/hl/stp/tactic/tactic.h (hunk 23 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/tactic.h:147]: If performance is a concern, it's probably worth profiling this, I suspect many of the checks will be optimized out, and even without that, the overhead is incredibly negligible with some branch prediction.

**Review by garethellis0 — COMMENTED** [review]
Nice, looking good!

File: src/software/ai/hl/stp/play_info.cpp (hunk 18 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play_info.cpp:4]: this is for displaying it in the visualizer in robot id order

File: src/software/ai/hl/stp/stp.cpp (hunk 20 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:33]: this is nothing new, just setting default values, so will keep it in mind but not in this pr

File: src/software/ai/hl/stp/tactic/move_tactic.h (hunk 21 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_tactic.h]: The FSM does need to be publicly exposed for other Tactics to use this FSM as a sub-FSM. I did consider putting just the FSM in a separate file, but I wasn't sure that'd be any better

File: src/software/ai/hl/stp/tactic/tactic.h (hunk 23 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic.h:147]: made https://github.com/UBC-Thunderbots/Software/issues/1911 to track this issue

File: src/software/ai/hl/stp/stp.cpp (hunk 20 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:33]: tbh it's not immediately obvious how to cleanly remember the previous play override

File: src/software/ai/hl/stp/play/play.h (hunk 17 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.h:96]: shouldn't this be obvious from the function definition: `std::vector<std::unique_ptr<Intent>> get`

File: src/software/ai/hl/stp/play/example_play_test.cpp (hunk 14 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/example_play_test.cpp:1]: These are the only non-sim play tests. I'm thinking about moving simplaytests to the play folder

File: src/software/ai/hl/stp/play/play.cpp (hunk 16 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.cpp:63]: What if, god forbid, we get a yellow card and we have 5 robots vs 6 tactics? Or we're field testing. I think it's too easy to get into situations where this warning will spam our logs.

File: src/software/ai/hl/stp/play/play.cpp (hunk 15 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.cpp:52]: done

File: src/software/ai/hl/stp/play/play.h (hunk 17 above)
Comment (jordanzliu) [reply src/software/ai/hl/stp/play/play.h:96]: this will end up confusing if someone assigns this to `auto`. `play.get()` doesn't tell me anything about what this function is supposed to do

Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.h:96]: I would hope people would name variables appropriately, i.e. `auto intents = play.get()`

**Comment by jonathanlew** [comment]
>     1. How to composite FSMs look in general? Are they are easy to follow as we think across multiple levels?

I will have an example of this up right after this PR goes in

**Comment by MathewMacDougall** [comment]
> > ```
> > 1. How to composite FSMs look in general? Are they are easy to follow as we think across multiple levels?
> > ```
> 
> I will have an example of this up right after this PR goes in

Isn't this PR part of the proposal though? Shouldn't we see what that looks like before merging this PR / accepting the proposal?

**Comment by jonathanlew** [comment]
> > > ```
> > > 1. How to composite FSMs look in general? Are they are easy to follow as we think across multiple levels?
> > > ```
> > 
> > 
> > I will have an example of this up right after this PR goes in
> 
> Isn't this PR part of the proposal though? Shouldn't we see what that looks like before merging this PR / accepting the proposal?

Well, the question here isn't whether we accept the proposal or not, because this is something that we're going ahead with; there were no major objections to the RFC and there are clear benefits from this new architecture. This is what a [composite FSM will look like](https://github.com/jonathanlew/UBC-Thunderbots-Software/compare/tactic_play_stp_interface...jonathanlew:composite_fsm_chip_tactic) imo it's relatively straightforward

**Comment by MathewMacDougall** [comment]
> > > > ```
> > > > 1. How to composite FSMs look in general? Are they are easy to follow as we think across multiple levels?
> > > > ```
> > > 
> > > 
> > > I will have an example of this up right after this PR goes in
> > 
> > 
> > Isn't this PR part of the proposal though? Shouldn't we see what that looks like before merging this PR / accepting the proposal?
> 
> Well, the question here isn't whether we accept the proposal or not, because this is something that we're going ahead with; there were no major objections to the RFC and there are clear benefits from this new architecture. This is what a [composite FSM will look like](https://github.com/jonathanlew/UBC-Thunderbots-Software/compare/tactic_play_stp_interface...jonathanlew:composite_fsm_chip_tactic) imo it's relatively straightforward

Ah didn't know that. Took a quick look and it seems reasonable. More curious what it will look like with 2 or more composite FMSs within another FSM that need to potentially switch back and forth. I do think it's overall a good idea so am interested to see how things go.

File: src/software/ai/hl/stp/tactic/move_fsm.h [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+        // The point the robot is trying to move to
+        Point destination;
+        // The orientation the robot should have when it arrives at its destination
+        Angle final_orientation;
+        // The speed the robot should have when it arrives at its destination
+        double final_speed;
+    };
+
+    // this struct defines the only event that the MoveFSM responds to
+    UPDATE_STRUCT_WITH_CONTROL_PARAMS_AND_COMMON
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/move_fsm.h]: nit: `UPDATE_STRUCT_WITH_CONTROL_PARAMS_AND_COMMON` -> `UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS` ?

File: src/software/ai/hl/stp/tactic/stop_fsm.h [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+struct StopFSM
+{
+    class idle_state;
+    class stop_state;
+
+    struct Update
+    {
+        Update(bool coast, const TacticUpdate& common) : coast(coast), common(common) {}
+        bool coast;
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/stop_fsm.h:13]: we should pop `coast` in a `ControlParams` struct and use the macro here

**Review by akhilveeraghanta — DISMISSED** [review]
LGTM :fire: just 2 nits

File: src/software/ai/hl/stp/tactic/stop_fsm.h (hunk 26 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/stop_fsm.h:13]: done

File: src/software/ai/hl/stp/tactic/move_fsm.h (hunk 25 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/move_fsm.h]: ok

File: src/software/ai/hl/stp/play/play.h (hunk 17 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/play/play.h:96]: If you're choosing between the two of them, might as well choose the more descriptive one no? If I'm using autocomplete on an instance of this class, I'm definitely going to have to go lookup the function signature to understand what this does.

File: src/software/ai/hl/stp/play_info.cpp (hunk 18 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/play_info.cpp:4]: If the only goal here is to change how the information is displayed in the visualizer, I'd really recommend changing `std::string` to something with more information (such as a struct with a robot id and tactic reference) so that the visualizer can be responsible for deciding how things are displayed, rather then something in STP making decisions about visualization.

**Review by itsarune — COMMENTED** [review]
👍 I think I understand the new framework, I'd also be interested in seeing what a more complicated tactic would look like

File: src/software/ai/hl/stp/play/play.h (hunk 17 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.h:96]: I don't understand the big fuss over this since we will only be calling this from STP. imo, it's cleaner to say the interface for play is to return intents and we provide a `get` function to get those intents. (It might not be clear now since we're supporting both interfaces simultaneously, but that's where we're going with this)

> If I'm using autocomplete on an instance of this class, I'm definitely going to have to go lookup the function signature to understand what this does.

Shouldn't we all at least look at the function definition before using it?

Following this logic of self-describing function, the logical conclusion would be calling it `getIntentUniquePtrVector`, no?

Comment (garethellis0) [reply src/software/ai/hl/stp/play/play.h:96]: I suppose the difference for me is I have no idea what the function called `get` does, but I at least know the semantics of what I'm getting if it's called `getIntents` (vs. the syntactics, which is what `getIntentUniquePtrVector` conveys). 

I'm not saying `get` is incorrect, I just think `getIntents` is a more semantically descriptive name that's more generally consistent with how getter functions are named in this codebase.

(All the above said, all my comments are really just intended as suggestions. It's absolutely your guys's codebase now, so please don't feel like you have to justify anything to me!)

File: src/software/ai/hl/stp/play_info.cpp (hunk 18 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play_info.cpp:4]: added to https://github.com/UBC-Thunderbots/Software/issues/1831

File: src/software/ai/hl/stp/play/play.h (hunk 17 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/play.h:96]: Well, there seem to be other people on the team that aren't convinced, so it seems worth explaining my thinking.
`getIntents` could return so many things: `std::set<std::weak_ptr<Intent>>`? `std::vector<std::shared_ptr<Intent>>`? `std::unorderered_set<Intent*>`? `std::vector<std::shared_ptr<const Intent>>`? 
I don't think we should be making these assumptions about how our codebase works to somehow deduce that `getIntents` means a vector of unique ptrs. Please please please read function signatures.

File: src/software/ai/hl/stp/tactic/tactic_fsm.h [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+/**
+ * The Update struct is the only event that a tactic fsm should respond to and it is
+ * composed of the following structs:
+ *
+ * ControlParams - uniquely defined by each tactic to control the FSM
+ * TacticUpdate - common struct that contains Robot, World, and SetIntentCallback
+ */
+#define UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS                                     \
```
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/tactic_fsm.h]: nit: this macro name sounds like it's updating a struct, instead of declaring an `Update` struct. Would be nice to add `DEFINE_` to this name

**Review by jordanzliu — COMMENTED** [review]
LGTM other than existing comments, great documentation on `move_tactic`

File: src/software/ai/hl/stp/tactic/tactic_fsm.h (hunk 27 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/tactic_fsm.h]: fixed

**Review by akhilveeraghanta — APPROVED** [review]

**Review by jordanzliu — APPROVED** [review]
LGTM 💯
